Remove server mode filter from platform parameter filters

lkbhitesh07 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: This PR removes the server mode filter from the platform parameters and feature flags. The filter is not needed because the server mode will be determined by the server we are editing the platform parameters on and the changes will work on that server.
3. (For bug-fixing PRs only) The original bug occurred because: N/A

**Note:** Please note that we have to follow a few steps both before and after the release, the instructions are mentioned in this [doc](https://docs.google.com/document/d/1w4zaCNyDR7mWJJ0M-K1LXHRBilWg__13x5KTN7dj2s8/edit?usp=sharing), I will also be submitting the testing form in order to include it as the actions need to be done with the release, I will make sure that @kevintab95 have the info regarding this.

## Essential Checklist

- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct

![Screenshot from 2023-06-29 16-10-48](https://github.com/oppia/oppia/assets/44300735/e74ca0ae-abdb-4a4a-a364-f17f11f2984c)

**The features works fine as per their feature stage without having server_mode filter**
Please note that for this I shifted the e2e_dummy_feature to dev feature stage, when we enable the feature we can see a small div at the bottom left that tells the feature is enabled and after shifting to prod mode we are no longer able to see it.

[Screencast from 04-07-23 01:43:58 PM IST.webm](https://github.com/oppia/oppia/assets/44300735/f9b2b0f2-912c-4191-a123-45e58a84646e)

**The default value is not editable on the features tab**
![Screenshot from 2023-07-07 13-48-51](https://github.com/oppia/oppia/assets/44300735/ba0780c6-2b17-470f-9082-cd4f94d7909f)


## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

lkbhitesh07 committed:
initial changes

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
fixed tests

lkbhitesh07 committed:
fixed e2e test

lkbhitesh07 committed:
fixed lint

lkbhitesh07 committed:
fixed the failing tests

lkbhitesh07 committed:
ts check fix

lkbhitesh07 committed:
fixed frontend tests

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
updates to the features tab

lkbhitesh07 committed:
bit

lkbhitesh07 committed:
changed feature stage of dummy-feature

lkbhitesh07 committed:
fixed backend test

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
added blank line to the spec file

lkbhitesh07 committed:
change the default value of feature flag

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into edit-default-value-when-no-rules

lkbhitesh07 committed:
added the default_value field to the model

lkbhitesh07 committed:
added for feature flags

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into edit-default-value-when-no-rules

Commit:
handled the incoming schema for feature flags

Commit:
fixed backend tests

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into edit-default-value-when-no-rules

Commit:
fixed tests

Commit:
fixed backend tests

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into edit-default-value-when-no-rules

Commit:
fixed tests

Commit:
fixed lint

Commit:
fixed single mypy

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into edit-default-value-when-no-rules

Commit:
removed server_mode filter from backend

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into edit-default-value-when-no-rules

Commit:
fixed mypy tests

Commit:
more mypy

Commit:
added button to update the default values

Commit:
fixed lint

Commit:
more lint

Commit:
nit

Commit:
return default value from parameter registry when not present in storage model

Commit:
added test

Commit:
fixed lint

Commit:
fixed mypy

Commit:
Merge branch 'edit-default-value-when-no-rules' of https://github.com/lkbhitesh07/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
nit

Commit:
Fixed the frontend

Commit:
nit

Commit:
nits for backend

Commit:
fixed lint

Commit:
fixed frontend tests

lkbhitesh07 commented:
Not assigning any reviewers now as this PR contains the changes of #18413, once that gets merged I will assign the apprpriate reviewers.
Thanks

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

lkbhitesh07 commented:
@seanlip @JayVivarekar @aasiffaizal @gp201 @Rijuta-s @sagangwee PTAL

seanlip commented on core/controllers/diagnostic_test_player.py:
```diff
@@ -43,7 +43,8 @@ class DiagnosticTestPlayerPage(
         if platform_feature_services.is_feature_enabled(
-            platform_feature_list.ParamNames.DIAGNOSTIC_TEST.value):
+            platform_feature_list.ParamNames.DIAGNOSTIC_TEST.value
+        ) and constants.DEV_MODE:
```
Wait, why is this here? I don't see why this should condition on dev mode, the is_feature_enabled() should take the server mode into account.

lkbhitesh07 replied:
I have removed it, I was doing some testing at that time.

seanlip commented on core/controllers/diagnostic_test_player_test.py:
```diff
@@ -44,18 +47,20 @@ def setUp(self) -> None:
                 })
             ],
-            False
+            True
```
Why change this?

lkbhitesh07 replied:
This is the default_value, so in case no rules are matched the default values are passed. I would require this as I want the value of the platform param on current server to be True.

seanlip commented on core/controllers/platform_feature_test.py:
```diff
@@ -249,97 +239,16 @@ def _mock_dummy_feature_stage(
             param_list.ParamNames.DUMMY_FEATURE_FLAG_FOR_E2E_TESTS.value)
         return self.swap(feature, '_feature_stage', stage.value)
 
-    def test_get_with_dummy_feature_enabled_in_dev_returns_ok(self) -> None:
```
There still need to be tests for dev and prod mode -- i.e. the server mode. The difference now is that dev/prod mode behaviour is determined based on the current "feature stage" of the flag and the current server mode, rather than an explicit "server mode" filter.

lkbhitesh07 replied:
Yes, that is correct. Please note that this feature is the e2e_dummy_feature which is in prod mode now and we just want to test the functionality here. As the feature is in prod mode this is available in all the server stage.
The tests related to the server modes and feature stages are present in the `feature_services_test` file. We can say that we were simply repeating the tests with e2e_dummy_feature in this file so I have removed this part.
Let me know what do you think.

seanlip replied:
I'm OK with anything as long as (a) we have full coverage of behaviours, (b) the test name correctly describes what the test is testing, (c) the test is the simplest test possible to test whatever it's testing.

seanlip commented on core/domain/learner_group_services_test.py:
```diff
@@ -140,16 +140,14 @@ def test_is_learner_group_feature_enabled(self) -> None:
                 })
             ],
-            False
+            True
```
Why is this changed?

lkbhitesh07 replied:
Same as [here](https://github.com/oppia/oppia/pull/18564#discussion_r1249176105)

seanlip replied:
But why has it got to be changed to True now when it wasn't before? Usually the default value for all feature flags should be False and they are manually set to True by the admin (based on a list of filters, possibly empty), so I think this might be incorrectly modelled.

In fact I think you should add a general backend test that confirms that the default value of each feature flag parameter is always False. Some of the code logic assumes this, e.g. see core/domain/platform_feature_services.py line 277 (after your edits).

lkbhitesh07 replied:
Got it, thanks for the explanation. I have made the changes and now the default_value at all the places is False.

I have added the backend test as well.

seanlip reviewed (changes requested):
@lkbhitesh07 I started taking a pass on this PR but I'm quite confused about it. Some general comments:

- The PR does a lot of changes relating to adding a platform_type filter. There is nothing at all about this in the PR description so I don't understand the rationale of what is going on and why this is being added. Could you please explain this more clearly?
- I think there needs to be some proof that simulates the local server to be in dev / test / prod mode and that shows the effect of the current server mode on feature flags that are in the dev / test / prod feature stages.

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
restored tests

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
fixed lint

Commit:
fix mypy

lkbhitesh07 commented:
> @lkbhitesh07 I started taking a pass on this PR but I'm quite confused about it. Some general comments:
> 
> * The PR does a lot of changes relating to adding a platform_type filter. There is nothing at all about this in the PR description so I don't understand the rationale of what is going on and why this is being added. Could you please explain this more clearly?
> * I think there needs to be some proof that simulates the local server to be in dev / test / prod mode and that shows the effect of the current server mode on feature flags that are in the dev / test / prod feature stages.

Hey @seanlip, thanks for the review.
Regarding the changes of adding the `platform_type` filter or any other type filter is just to replace the server_mode filter, we could also remove all the rules and just have an empty list. Basically, whenever we were creating any platform-params in the codebase instead of keeping the rules empty we would use server_mode filter and now as we are deprecating it I replaced that filter with some other flter.
We can always remove the filters and just have empty rules for platform params.

Regarding the proof you requested, I have added in the description.

EDIT: I have also updated the PR description a bit.
Thanks

lkbhitesh07 commented:
Done with the changes @seanlip.
Thanks

Commit:
added some more tests

Commit:
lint fix

seanlip commented on core/controllers/diagnostic_test_player_test.py:
```diff
@@ -36,40 +35,41 @@ def setUp(self) -> None:
-                feconf.DIAGNOSTIC_TEST_PLAYER_PAGE_URL,
-                expected_status_int=404
-            )
+        feature_services.update_feature_flag(
```
This test (and quite a few of the other tests you've edited) should have a pre-check.

lkbhitesh07 replied:
I have added the precheck to the tests I have added.

seanlip commented on core/controllers/platform_feature_test.py:
```diff
@@ -217,22 +214,15 @@ def tearDown(self) -> None:
-        self, is_enabled: bool, mode: param_domain.ServerMode
-    ) -> None:
+    def _set_dummy_feature_status(
+        self, is_enabled: bool) -> None:
```
feature_is_enabled (since it's a bool and not a function, don't start with a verb)

lkbhitesh07 replied:
Done

seanlip commented on core/controllers/platform_feature_test.py:
```diff
@@ -249,97 +239,16 @@ def _mock_dummy_feature_stage(
-                expected_status_int=404
-            )
+    def test_get_with_dummy_feature_disabled_raises_404(self) -> None:
+        self._set_dummy_feature_status(False)
```
Here's another example of a missing pre-check. Please check the other test files as well.

lkbhitesh07 replied:
I have added the prechecks.

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -270,9 +270,13 @@ def _evaluate_feature_flag_values_for_context(
             feature_name)
-        feature_name_value = param.evaluate(context)
+        assert feature.feature_stage is not None
+        if not feature_flag_valid_on_current_server(feature.feature_stage):
```
Can't this logic be part of feature.evaluate()? (Presumably it should be the very first thing we check there.)

lkbhitesh07 replied:
I have added it there.

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -321,3 +325,25 @@ def get_platform_parameter_value(
     return param.evaluate(context)
+
+
+def feature_flag_valid_on_current_server(feature_stage: str) -> bool:
```
This is a function so its name should start with a verb.

lkbhitesh07 replied:
No longer valid as the logic has been moved.

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -321,3 +325,25 @@ def get_platform_parameter_value(
+        feature_stage: str. The stage of the feature flag.
+
+    Returns:
+        bool. Returns True if the feature is valid on current server else False.
```
bool. Whether the feature is valid on the current server.

lkbhitesh07 replied:
No longer valid as the logic has been moved.

seanlip commented on core/domain/platform_feature_services_test.py:
```diff
@@ -92,16 +93,14 @@ def setUp(self) -> None:
                 })
             ],
-            False
+            True
```
Same notes here and elsewhere, the default value should be False for feature flags and there should be a backend test verifying that (as well as potentially validation in the domain layer -- or in fact, don't even allow it to be passed into the constructor for feature flags).

This convention is arguable so let me also try to explain my reasoning. I'm concerned that, with a default value of True, it will reasonably be assumed that such flags would also have a value of True in servers where they are not valid for their feature stage (since, the default value is the value that flags "reset" to). So I think it would be better to model this as a default value of False and a no-filter rule that changes the value to True, since rules can only be set in the relevant feature stage. What do you think?

lkbhitesh07 replied:
Thanks for the detailed explanation, I agree with your reasoning. I have made the changes.
I have added the backend test as requested and also wanted to let you know that even now we do not allow the default_value to be passed in the feature flags constructor.
I'm a bit unsure about adding the validation in the domain layer, as per the recent changes we are allowed to make changes to the default_value of feature-flags and platform-params, adding the validation will result in failing for that functionality.

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -270,9 +270,13 @@ def _evaluate_feature_flag_values_for_context(
+        feature = registry.Registry.get_platform_parameter(
             feature_name)
-        feature_name_value = param.evaluate(context)
+        assert feature.feature_stage is not None
```
Why is this assertion needed?

lkbhitesh07 replied:
Not relevant anymore, removed after new changes.

seanlip commented on core/domain/platform_parameter_domain.py:
```diff
@@ -911,34 +885,38 @@ def _validate_feature_flag(self) -> None:
+            self._feature_stage == ServerMode.DEV.value
+        ):
+            raise utils.ValidationError(
+                'Feature in %s stage cannot be enabled in %s environment.' % (
```
What about this check suggests "enabled"? Is it more like it cannot be updated?

Ditto below.

lkbhitesh07 replied:
Updated.

seanlip commented on core/domain/platform_parameter_registry_test.py:
```diff
@@ -247,92 +248,49 @@ def test_update_parameter_with_invalid_rules_failure(self) -> None:
-                False
-            )
-
-    def test_update_test_feature_with_rule_enabled_for_prod_raises_exception(
```
The corresponding test for this seems missing.

lkbhitesh07 replied:
I have added it.

seanlip reviewed (changes requested):
Thanks, reviewed codeowner files.

> Regarding the changes of adding the platform_type filter or any other type filter is just to replace the server_mode filter, we could also remove all the rules and just have an empty list. Basically, whenever we were creating any platform-params in the codebase instead of keeping the rules empty we would use server_mode filter and now as we are deprecating it I replaced that filter with some other flter. We can always remove the filters and just have empty rules for platform params.

Thanks for the explanation. Let's follow the guideline of having the tests be the simplest possible ones to test whatever behaviour they're supposed to be testing (though in a string of tests that test similar non-filter-related stuff you could put in some occasional filters to vary things up a bit). So, if filters are important to the test in question then let's keep them, but if they're not I see no problem with having zero filters.

Also one question -- what's the plan for migrating existing parameters which have a server mode dependency? It's fine if this is done manually but has this been included in the release instructions? Presumably some things need to be done (to reset the server_mode filter) just before the release too -- but there's nothing in the PR description about this. /cc @kevintab95

Commit:
done suggested changes

Commit:
fixed lint

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
nit

Commit:
added backend test

lkbhitesh07 commented:
> Thanks for the explanation. Let's follow the guideline of having the tests be the simplest possible ones to test whatever behaviour they're supposed to be testing (though in a string of tests that test similar non-filter-related stuff you could put in some occasional filters to vary things up a bit). So, if filters are important to the test in question then let's keep them, but if they're not I see no problem with having zero filters.

Got it, I have made the changes.


> Also one question -- what's the plan for migrating existing parameters which have a server mode dependency? It's fine if this is done manually but has this been included in the release instructions? Presumably some things need to be done (to reset the server_mode filter) just before the release too -- but there's nothing in the PR description about this. /cc @kevintab95

Thanks for mentioning it, I have a question regarding this.
I'm a bit unsure regarding the steps we need to follow in order to reset the server_mode filter before the release, Initially I thought that we can simply create a filter with no conditions in order to turn that feature on but we cannot do that as currently we have a backend validation that says `All rules must have a server_mode filter.`. This PR removes that validation.
What should we do in this case?
I will update the PR description as soon as this part will be clear.

Thanks

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -270,13 +270,9 @@ def _evaluate_feature_flag_values_for_context(
-            result_dict[feature_name] = False
-            continue
-        feature_name_value = feature.evaluate(context)
+        feature_name_value = param.evaluate(context)
```
Why do we call this feature_name_value rather than feature_value (or feature_is_enabled)? What is the "value" of a "feature name"?

lkbhitesh07 replied:
Done

seanlip reviewed (changes requested):
This seems pretty good. Thanks @lkbhitesh07! Just one inline comment.

Also for the more general points:

> I'm a bit unsure about adding the validation in the domain layer, as per the recent changes we are allowed to make changes to the default_value of feature-flags and platform-params, adding the validation will result in failing for that functionality. 

I think we should prohibit default_values being edited for feature flags (either in the frontend or the backend). It's fine to allow them to be edited for platform parameters. We should make sure that such a change doesn't get through to the backend and also use validate() to guard the storage layer against such a change.

>> Also one question -- what's the plan for migrating existing parameters which have a server mode dependency? It's fine if this is done manually but has this been included in the release instructions? Presumably some things need to be done (to reset the server_mode filter) just before the release too -- but there's nothing in the PR description about this.  /cc @kevintab95 
>
> I'm a bit unsure regarding the steps we need to follow in order to reset the server_mode filter before the release, Initially I thought that we can simply create a filter with no conditions in order to turn that feature on but we cannot do that as currently we have a backend validation that says All rules must have a server_mode filter. This PR removes that validation. What should we do in this case?

Hm, that's interesting. Maybe we can just temporarily switch that filter to use something like platform_type instead, prior to deployment?

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
renamed the variable

lkbhitesh07 commented:
Thanks for the review @seanlip,

> I think we should prohibit default_values being edited for feature flags (either in the frontend or the backend). It's fine to allow them to be edited for platform parameters. We should make sure that such a change doesn't get through to the backend and also use validate() to guard the storage layer against such a change.

Actually, recently we have introduced a new `default_value` field for the platform storage model(relevant PR #18413) and in order to populate that we make a request to the backend to update the feature flag default values, so we would require that in the backend.
I can maybe disable the editing of the default value from the frontend for feature flags.
Well come to think of it there is a turnaround that we can do, we can avoid sending the default value from the frontend and we can manage to get it in the backend.
Let me know what do you think.


> Hm, that's interesting. Maybe we can just temporarily switch that filter to use something like platform_type instead, prior to deployment?

Sounds good, I have updated the PR description with the instructions.
Thanks

seanlip commented:
> Actually, recently we have introduced a new default_value field for the platform storage model(relevant PR #18413) and in order to populate that we make a request to the backend to update the feature flag default values, so we would require that in the backend. I can maybe disable the editing of the default value from the frontend for feature flags. Well come to think of it there is a turnaround that we can do, we can avoid sending the default value from the frontend and we can manage to get it in the backend. Let me know what do you think.

Yup, I think (a) the frontend should not allow editing the default value for feature flags, (b) there needs to be a check in validate() to prevent default values being set to anything but False if the platform parameter is a feature flag. You cannot do without (b) since this will lead to potential corruption on the server.

> Sounds good, I have updated the PR description with the instructions.

Thanks, I left some comments in the doc.

aasiffaizal commented on core/domain/platform_parameter_domain.py:
```diff
@@ -341,19 +339,14 @@ def _evaluate_single_value(
-            self._type in ['server_mode', 'platform_type', 'browser_type']
-            and op != '='
-        ):
+        if self._type in ['platform_type', 'browser_type'] and op != '=':
```
```suggestion
        if self._type in {'platform_type', 'browser_type'} and op != '=':
```

lkbhitesh07 replied:
Done

aasiffaizal commented on core/domain/platform_parameter_domain.py:
```diff
@@ -879,6 +864,19 @@ def evaluate(
         """
         if context.is_valid:
+            if self._is_feature:
+                server_mode = self._get_server_mode().value
+                if (
+                    server_mode == ServerMode.TEST.value and
+                    self._feature_stage == ServerMode.DEV.value
+                ):
+                    return False
+                if (
+                    server_mode == ServerMode.PROD.value and
+                    self._feature_stage in (
+                        ServerMode.DEV.value, ServerMode.TEST.value)
+                ):
+                    return False
```
```suggestion
                server_mode = self._get_server_mode()
                if (
                    server_mode == ServerMode.TEST and
                    self._feature_stage == ServerMode.DEV
                ):
                    return False
                if (
                    server_mode == ServerMode.PROD and
                    self._feature_stage in {
                        ServerMode.DEV, ServerMode.TEST
                    }
                ):
                    return False
```

Shouldn't this work?

aasiffaizal replied:
ditto, for the same occurrence below

lkbhitesh07 replied:
It will not work because the `self._feature_stage` has type of str and we would require to get the value of server mode enum. I have made the changes for the `server_mode` part.

aasiffaizal reviewed (changes requested):
Thanks @lkbhitesh07, Took a pass and left a few comments.

Commit:
remove the editing of feature flag default value and added the validation in domain layer

Commit:
fixed lint

Commit:
fixed frontend tests

lkbhitesh07 commented:
Thanks @seanlip and @aasiffaizal, I have made the changes.
Thanks

gp201 reviewed (approved):
LGTM for codeowner files

seanlip commented on core/domain/platform_parameter_registry.py:
```diff
@@ -194,7 +194,8 @@ def update_platform_parameter(
         commit_message: str,
         new_rules: List[platform_parameter_domain.PlatformParameterRule],
-        default_value: platform_parameter_domain.PlatformDataTypes
+        default_value: Optional[
```
This should not be optional. It should always be specified. It's just that, when calling this method from the method "just above" it which uses a feature flag, you should pass False explicitly and give an explanation in a comment next to that code (i.e. that the default value of a feature flag is always False).

Try to avoid default parameters in general, since it ends up requiring branching (like you have later on).

lkbhitesh07 replied:
Done

seanlip commented on core/domain/platform_parameter_domain.py:
```diff
@@ -339,7 +339,7 @@ def _evaluate_single_value(
             Exception. Given operator is not supported.
         """
-        if self._type in ['platform_type', 'browser_type'] and op != '=':
+        if self._type in {'platform_type', 'browser_type'} and op != '=':
```
It doesn't matter too much, but typically we use tuples (..., ...) for this sort of thing.

lkbhitesh07 replied:
Updated

JayVivarekar reviewed (approved):
Thanks @lkbhitesh07! LGTM for codeowner files.

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
nits

kevintab95 reviewed (changes requested):
Adding a review hold until this gets tested on the server.

seanlip commented on core/domain/platform_feature_services.py:
```diff
@@ -194,8 +194,11 @@ def update_feature_flag(
     registry.Registry.update_platform_parameter(
-        feature_name, committer_id, commit_message, new_rules)
+        feature_name, committer_id, commit_message, new_rules,
+        default_value=False)
```
Since this is not an optional parameter, you don't need `default_value=` here.

lkbhitesh07 replied:
Done

seanlip reviewed (approved):
@lkbhitesh07 LGTM, just one minor comment.

seanlip commented on core/domain/platform_parameter_registry.py:
```diff
@@ -213,8 +212,6 @@ def update_platform_parameter(
         # Create a temporary param instance with new rules for validation,
         # if the new rules are invalid, an exception will be raised in
         # validate() method.
```
Just to check, this will validate that the default value has the correct type for the given platform parameter?

lkbhitesh07 replied:
Yes, that is correct.

Rijuta-s reviewed (approved):
Looks Good to me for code owners!

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
nit

lkbhitesh07 commented:
Hey @seanlip, @kevintab95
Do we require to test this on the server before merging it? We are simply trying to replace the "server_mode" filter with the "platform_type" filter and I think it would be fine because we are using the "platform_type" filter property for a while now and it seems to work completely fine.
Let me know what you guys think?

Thanks

seanlip commented:
@lkbhitesh07 I don't have a strong opinion on this. It's something you might want to discuss the pros/cons of with the release team and server admins, since they are responsible for the integrity of the production servers.

aasiffaizal reviewed (approved):
LGTM

kevintab95 reviewed (approved):
Hi @lkbhitesh07, on further review, I think it is OK to merge this one as we are simply dropping an existing filter option. Some manual steps need to be performed before this code is deployed -- your entry using the job form will serve as a reminder for the release coordinator to perform the said steps. So thank you for submitting the request. LGTM!

kevintab95 reviewed (changes requested):
Blocking this since updating the flag on the server failed. It needs to be sorted out before merging the PR.

lkbhitesh07 commented:
Thanks @kevintab95, For now I have removed this PR from the merge queue as I recently found out that filters other than "server_mode" does not work properly for the platform params and the steps mentioned in the job-form doc involve adding a filter of type `platform_type`. Will be raising a PR regarding that today and that will be the prerequisite for this PR.

Thanks

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
fixed some tests

Commit:
fixed lint

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
fixed nit

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-server-mode-filter

JayVivarekar reviewed (approved):
LGTM! Tested successfully on the backup server.

lkbhitesh07 commented:
Hey @kevintab95 the PR has been successfully tested on the server, I'm waiting on your approval.

Thanks

lkbhitesh07 committed:
Merge branch 'develop' into remove-server-mode-filter

Commit:
fixed tests
